File: output/typescript/tests/03_transition.ts
class WithTransitionFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class WithTransitionFrameContext {
    public event: WithTransitionFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: WithTransitionFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class WithTransitionCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: WithTransitionCompartment | null;

    constructor(state: string, parent_compartment: WithTransitionCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): WithTransitionCompartment {
        const c = new WithTransitionCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class WithTransition {
    private _state_stack: Array<any>;
    private __compartment: WithTransitionCompartment;
    private __next_compartment: WithTransitionCompartment | null;
    private _context_stack: Array<any>;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new WithTransitionCompartment("First");
        this.__next_compartment = null;
        const __frame_event = new WithTransitionFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: WithTransitionFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new WithTransitionFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new WithTransitionFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new WithTransitionFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: WithTransitionFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: WithTransitionCompartment) {
        this.__next_compartment = next_compartment;
    }

    public next() {
        const __e = new WithTransitionFrameEvent("next", null);
        const __ctx = new WithTransitionFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    public get_state(): string {
        const __e = new WithTransitionFrameEvent("get_state", null);
        const __ctx = new WithTransitionFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    private _state_First(__e: WithTransitionFrameEvent) {
        if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "First";
            return;;
        } else if (__e._message === "next") {
            console.log("Transitioning: First -> Second");
            const __compartment = new WithTransitionCompartment("Second", this.__compartment.copy());
            this.__transition(__compartment);
        }
    }

    private _state_Second(__e: WithTransitionFrameEvent) {
        if (__e._message === "get_state") {
            this._context_stack[this._context_stack.length - 1]._return = "Second";
            return;;
        } else if (__e._message === "next") {
            console.log("Transitioning: Second -> First");
            const __compartment = new WithTransitionCompartment("First", this.__compartment.copy());
            this.__transition(__compartment);
        }
    }
}


function main() {
    console.log("=== Test 03: State Transitions ===");
    const s = new WithTransition();

    // Initial state should be First
    let state = s.get_state();
    if (state !== "First") {
        throw new Error(`Expected 'First', got '${state}'`);
    }
    console.log(`Initial state: ${state}`);

    // Transition to Second
    s.next();
    state = s.get_state();
    if (state !== "Second") {
        throw new Error(`Expected 'Second', got '${state}'`);
    }
    console.log(`After first next(): ${state}`);

    // Transition back to First
    s.next();
    state = s.get_state();
    if (state !== "First") {
        throw new Error(`Expected 'First', got '${state}'`);
    }
    console.log(`After second next(): ${state}`);

    console.log("PASS: State transitions work correctly");
}

main();

File: output/python/tests/03_transition.py
from typing import Any, Optional, List, Dict, Callable

class WithTransitionFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class WithTransitionFrameContext:
    def __init__(self, event: WithTransitionFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class WithTransitionCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'WithTransitionCompartment':
        c = WithTransitionCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class WithTransition:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.__compartment = WithTransitionCompartment("First")
        self.__next_compartment = None
        __frame_event = WithTransitionFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = WithTransitionFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = WithTransitionFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = WithTransitionFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def next(self):
        __e = WithTransitionFrameEvent("next", None)
        __ctx = WithTransitionFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def get_state(self) -> str:
        __e = WithTransitionFrameEvent("get_state", None)
        __ctx = WithTransitionFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_First(self, __e):
        if __e._message == "get_state":
            self._context_stack[-1]._return = "First"
            return
        elif __e._message == "next":
            print("Transitioning: First -> Second")
            __compartment = WithTransitionCompartment("Second", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def _state_Second(self, __e):
        if __e._message == "get_state":
            self._context_stack[-1]._return = "Second"
            return
        elif __e._message == "next":
            print("Transitioning: Second -> First")
            __compartment = WithTransitionCompartment("First", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)


def main():
    print("=== Test 03: State Transitions ===")
    s = WithTransition()

    # Initial state should be First
    state = s.get_state()
    assert state == "First", f"Expected 'First', got '{state}'"
    print(f"Initial state: {state}")

    # Transition to Second
    s.next()
    state = s.get_state()
    assert state == "Second", f"Expected 'Second', got '{state}'"
    print(f"After first next(): {state}")

    # Transition back to First
    s.next()
    state = s.get_state()
    assert state == "First", f"Expected 'First', got '{state}'"
    print(f"After second next(): {state}")

    print("PASS: State transitions work correctly")

if __name__ == '__main__':
    main()
